docs/src/coding-guidelines.md

```diff
@@ -0,0 +1,170 @@
+```
+over:
+```c++
+#include "something.h"
+```
+
+Using TTIRToTTNN.cpp as an example, this is what includes would look like for us:
+
+```
+#include "ttmlir/Conversion/TTIRToTTNN/TTIRToTTNN.h"  # main header
+
+#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"  # these are local/private headers
+#include "ttmlir/Dialect/TTNN/Utils/Utils.h"
+
+#include "mlir/Dialect/MemRef/IR/MemRef.h"  # llvm project/subproj headers
+#include "llvm/Support/LogicalResult.h"
+
+#include <cstdio>  # system includes
+#include <algorithm>
+```
+# Comments
+
+Write comments as full sentences, starting with a capital letter and ending with a period. Comments should explain why the code exists, not just what it does. Use comments to clarify logic, assumptions, or any non-obvious aspects of the code.
+
+Example of a comment:
+```c++
+// Initialize the buffer to store incoming data from the network.
+```
+
+In general use use C++-style comments (//). Use C-style comments (/**/) only for when documenting the significance of constants used as actual parameters in a call:
+```c++
+object.callFunction(/*arg0=*/nullptr);
+```
+
+Every function, class, or non-trivial piece of logic should have a comment. Avoid redundant comments for self-explanatory code, but never leave complex code unexplained. Example of redundant comment:
+```c++
+// Increment the counter by 1.  // Redundant, avoid.
+counter++;
+```
+
+Ensure comments are accurate and reflect the current state of the code. Outdated or misleading comments can be worse than no comments at all.
+
+# Code Denesting (Inversion)
+
+Strive to minimize unnecessary indentation without compromising code clarity. One effective way to achieve this is by using early exits and the continue keyword in long loops.
+
+Consider following example:
+```c++
+void doSomething(Operation *op)
+{
+    if (op->getNumOperands() > 0
+        && isDpsOp(op)
+        && doSomethingDifferent(op))
+    {
+        // ... some long code ...
+    }
+}
+```
+
+It is strongly recommended to format the code as follows:
+```c++
+void doSomething(Operation *op)
+{
+    // ...
+    // We need to do something with the op that has more than 0 operands
+    if (op->getNumOperands() <= 0 ) return;
+
+    // We need something to do with the DPS op
+    if (!isDpsOp(op)) return;
+
+    // Just for example purposes
+    if (!doSomethingDifferent(op)) return;
+
+    // .. some long code ...
+}
+```
+
+This reduces loop nesting, makes the reasoning behind the conditions clearer, and signals to the reader that there is no subsequent else to worry about, reducing cognitive load. This can significantly improve code readability and comprehension.
+
+# Helper Functions
+
+This coding guideline addresses visibility and linkage of simple helper functions to ensure clarity, prevent linking errors, and improve maintainability:
+1. If a helper function needs to be defined in a .cpp file, it should be declared **static** or wrapped inside an **anonymous namespace**.
```

Comment:
One big grief that I have with such functions is that they cannot be unit tested in isolation. I would recommend against doing this for any code that we might want to unit test (which is most of the code).